internal/docstore/docstore.go

```diff
@@ -179,21 +195,44 @@ func (l *ActionList) Update(doc Document, mods Mods) *ActionList {
 // See ActionList.Update.
 type Mods map[FieldPath]interface{}
 
-// Do executes the action list. If all the actions executed successfully, Do returns
-// (number of actions, nil). If any failed, Do returns the number of successful
-// actions and an error. In general there is no way to know which actions succeeded,
-// but the error will contain as much information as possible about the failures.
-func (l *ActionList) Do(ctx context.Context) (int, error) {
+// An ActionListError is returned by ActionList.Do. It contains all the errors
+// encountered while executing the ActionList, and the positions of the corresponding
+// actions.
+type ActionListError = driver.ActionListError
+
+// Do executes the action list.
+// In ordered mode (when the Unordered method was not called on the list), execution
+// will stop after the first action that fails.
+// In unordered mode, all the actions will be executed. In either mode, as a special
+// case, none of the actions will be executed if any is invalid (for example, a Put
+// whose document is missing its key field).
+// If any action fails, the returned error will be an ActionListError that contains
+// the position in the ActionList of each failed action.
+func (l *ActionList) Do(ctx context.Context) error {
 	var das []*driver.Action
-	for _, a := range l.actions {
+	var alerr ActionListError
+	for i, a := range l.actions {
 		d, err := a.toDriverAction()
 		if err != nil {
-			return 0, wrapError(l.coll.driver, err)
+			alerr = append(alerr, struct {
+				Index int
+				Err   error
+			}{i, wrapError(l.coll.driver, err)})
+		} else {
+			das = append(das, d)
 		}
-		das = append(das, d)
 	}
-	n, err := l.coll.driver.RunActions(ctx, das)
-	return n, wrapError(l.coll.driver, err)
+	if len(alerr) > 0 {
+		return alerr
+	}
+	alerr = l.coll.driver.RunActions(ctx, das, l.unordered)
+	if alerr != nil { // Explicit non-nil check is necessary because alerr is not of type error.
```

Comment:
Why is the nil check needed? Isn't it always safe to `range` over the slice?

Reply:
It is safe to range over a nil slice. But I need the explicit nil check somewhere. If there is no error, I want to return the interface value `(nil, nil)` from this function, not `(nil, ActionListError)`. See https://golang.org/doc/faq#nil_error.

My comment on this line was supposed to explain this. Let me know how I can improve it.

Reply:
Oh I see. I was misled by the comment being right before the `range`.

Maybe:

````
alerr = ...
if len(alerr) == 0 {
  return nil // Explicitly return nil, because alerr is not of type error
}
for i := range ...{}
return alerr
```

I don't feel that strongly about this, but I didn't get the comment on first read.